File: livedisplay/aidl/sysfs/sunlight_enhancement.rs
/*
 * SPDX-FileCopyrightText: 2025 The LineageOS Project
 * SPDX-License-Identifier: Apache-2.0
 */

use binder::{Interface, Result as BinderResult, StatusCode};
use std::fs;

use vendor_lineage_livedisplay::aidl::vendor::lineage::livedisplay::ISunlightEnhancement::ISunlightEnhancement;

use crate::livedisplay_utils;

const FILE_HBM: &str = "/sys/class/graphics/fb0/hbm";
const FILE_SRE: &str = "/sys/class/graphics/fb0/sre";

pub struct SunlightEnhancement {
    file: Option<String>,
    enabled_mode: i32,
}

impl SunlightEnhancement {
    pub fn new() -> Self {
        if livedisplay_utils::is_path_accessible(FILE_HBM) {
            Self {
                file: Some(FILE_HBM.to_string()),
                enabled_mode: 1,
            }
        } else if livedisplay_utils::is_path_accessible(FILE_SRE) {
            Self {
                file: Some(FILE_SRE.to_string()),
                enabled_mode: 2,
            }
        } else {
            Self {
                file: None,
                enabled_mode: 0,
            }
        }
    }

    pub fn is_supported(&self) -> bool {
        self.file.is_some()
    }
}

impl Interface for SunlightEnhancement {}

impl ISunlightEnhancement for SunlightEnhancement {
    fn getEnabled(&self) -> BinderResult<bool> {
        fs::read_to_string(self.file.clone().unwrap())
            .map_err(|err| {
                log::error!(
                    "Failed to read from {}: {}",
                    self.file.clone().unwrap(),
                    err
                );
                StatusCode::UNKNOWN_ERROR.into()
            })
            .and_then(|s| {
                s.trim()
                    .trim_end_matches('\0')
                    .parse::<i32>()
                    .map(|value| value > 0)
                    .map_err(|err| {
                        log::error!("Failed to parse number: {}", err);
                        StatusCode::UNKNOWN_ERROR.into()
                    })
            })
    }

    fn setEnabled(&self, enabled: bool) -> BinderResult<()> {
        fs::write(
            self.file.clone().unwrap(),
            if enabled {
                self.enabled_mode.to_string()
            } else {
                "0".to_string()
            },
        )
        .map_err(|err| {
            log::error!("Failed to write to {}: {}", self.file.clone().unwrap(), err);
            StatusCode::UNKNOWN_ERROR.into()
        })
    }
}

File: livedisplay/aidl/sysfs/SunlightEnhancement.cpp
/*
 * SPDX-FileCopyrightText: The LineageOS Project
 * SPDX-License-Identifier: Apache-2.0
 */

#define LOG_TAG "SunlightEnhancementService"

#include <livedisplay/sysfs/SunlightEnhancement.h>

#include <android-base/file.h>
#include <android-base/logging.h>
#include <android-base/strings.h>

using ::android::base::ReadFileToString;
using ::android::base::Trim;
using ::android::base::WriteStringToFile;

namespace aidl {
namespace vendor {
namespace lineage {
namespace livedisplay {
namespace sysfs {

SunlightEnhancement::SunlightEnhancement()
    : SunlightEnhancement([]() {
          constexpr const char* kHBMPaths[] = {
#ifdef SE_PATH
                  SE_PATH,
#else
                  "/sys/class/graphics/fb0/hbm",
                  "/sys/devices/platform/soc/soc:qcom,dsi-display-primary/dsi_display_hbm",
                  "/sys/devices/platform/soc/soc:qcom,dsi-display-primary/hbm",
#endif
          };
          constexpr const char* kSREPaths[] = {
#ifdef SRE_PATH
                  SRE_PATH,
#else
                  "/sys/class/graphics/fb0/sre",
#endif
          };

          for (const auto& path : kHBMPaths) {
              if (!access(path, R_OK | W_OK)) {
                  return std::make_pair(path, SeMode::HBM);
              }
          }

          for (const auto& path : kSREPaths) {
              if (!access(path, R_OK | W_OK)) {
                  return std::make_pair(path, SeMode::SRE);
              }
          }

          LOG(FATAL) << "None of SunlightEnhancement paths could be opened";

          return std::make_pair("", SeMode::NONE);
      }()) {}

ndk::ScopedAStatus SunlightEnhancement::getEnabled(bool* _aidl_return) {
    std::string tmp;
    if (!ReadFileToString(this->file_, &tmp, true)) {
        *_aidl_return = false;
        return ndk::ScopedAStatus::fromExceptionCode(EX_SERVICE_SPECIFIC);
    }

    *_aidl_return = std::stoi(Trim(tmp)) > 0;
    return ndk::ScopedAStatus::ok();
}

ndk::ScopedAStatus SunlightEnhancement::setEnabled(bool enabled) {
    if (!WriteStringToFile(enabled ? std::to_string(static_cast<int32_t>(enabled_mode_)) : "0",
                           file_, true)) {
        LOG(ERROR) << "Failed to set SunlightEnhancement state";
        return ndk::ScopedAStatus::fromExceptionCode(EX_SERVICE_SPECIFIC);
    }

    return ndk::ScopedAStatus::ok();
}

}  // namespace sysfs
}  // namespace livedisplay
}  // namespace lineage
}  // namespace vendor
}  // namespace aidl
